### ISSUE TITLE
V1.27.1 crashes on launch

### ISSUE THREAD

**Issue by arnep-de** [markusfisch/BinaryEye#114]
Device: Galaxy S4 Mini Plus
Android version: 10
LineageOS version: 17.1
App version: 1.27.1 (f-droid build)

BinaryEye crashes 1-2 second after launch. Camera is opened in "Code scan" mode, then immediate crash.

adb catlog says:
```
05-14 12:01:52.016  4208  4300 E AndroidRuntime: FATAL EXCEPTION: CameraCallbackHandlerThread
05-14 12:01:52.016  4208  4300 E AndroidRuntime: Process: de.markusfisch.android.binaryeye, PID: 4208
05-14 12:01:52.016  4208  4300 E AndroidRuntime: android.support.v8.renderscript.RSRuntimeException: Loading of ScriptC script failed.
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at android.support.v8.renderscript.ScriptC.<init>(Unknown Source:31)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at d.a.a.a.g.a.<init>(Unknown Source:10)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at d.a.a.a.h.a.<init>(Unknown Source:26)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at de.markusfisch.android.binaryeye.activity.CameraActivity.a(Unknown Source:2)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at de.markusfisch.android.binaryeye.activity.CameraActivity.a(Unknown Source:13)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at de.markusfisch.android.binaryeye.activity.CameraActivity.a(Unknown Source:0)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at de.markusfisch.android.binaryeye.activity.CameraActivity$c$a.onPreviewFrame(Unknown Source:16)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at android.hardware.Camera$EventHandler.handleMessage(Camera.java:1251)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
05-14 12:01:52.016  4208  4300 E AndroidRuntime: 	at android.os.HandlerThread.run(HandlerThread.java:67)
05-14 12:01:52.028   696  4519 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
05-14 12:01:52.032   696   714 W ActivityTaskManager:   Force finishing activity de.markusfisch.android.binaryeye/.activity.CameraActivity
05-14 12:01:52.132   696  1777 I DropBoxManagerService: add tag=event_log isTagEnabled=true flags=0x2
05-14 12:01:52.386   696  3077 I WindowManager: WIN DEATH: Window{c76b037 u0 de.markusfisch.android.binaryeye/de.markusfisch.android.binaryeye.activity.CameraActivity}
05-14 12:01:52.391   696   989 I ActivityManager: Process de.markusfisch.android.binaryeye (pid 4208) has died: vis+99 TOP 
```

**Comment by jjakob** [comment]
Can confirm, both 1.27.0 and 1.27.1 crash, even when opening it multiple times. The last working version is 1.26.0

**Comment by markusfisch** [comment]
TL;DR
Stay with me, I'm working on a fix!

Sorry for the inconvenience 😞 And thanks a lot for reporting this! 👍

Since you both run LineageOS, I think I've got a hunch what's causing the issue 🤔 Some time ago I found out that RenderScript only works in compat mode on LineageOS, so I (kind of) solved the problem by calling `RenderScript.forceCompat()` by default.

Unfortunately, `RenderScript.forceCompat()` means that the processing (of the camera frame) is done in software and not in hardware (with better performance). Because all of the devices I got my hands on supported running RenderScript natively, I assumed it must be some weird bug in Lineage.

So, to improve performance for everybody else, I chose to run `RenderScript.forceCompat()` only for LineageOS in 1.27.0 here ea0212c58eb9d50726a2db887c15352c77b5a0b0.

A couple of days later I learned it wasn't just Lineage, see https://github.com/markusfisch/BinaryEye/issues/113

In search for a compromise that would work for all systems, 1.27.1 does now run `RenderScript.forceCompat()` as soon as RenderScript does crash (because it requires compat mode) here 307401087e9439087c58bdceb575811d314a512b.

But because I *always* run `forceCompat()` on LineageOS no matter what the flag is (see https://github.com/markusfisch/BinaryEye/blob/307401087e9439087c58bdceb575811d314a512b/app/src/main/kotlin/de/markusfisch/android/binaryeye/app/BinaryEyeApp.kt#L17), I suppose the app keeps crashing on your devices because `forceCompat()` isn't required for your systems somehow 🤔

**Comment by jjakob** [comment]
There are more errors on mine:
```
2020-05-14 02:35:02,VERBOSE,RenderScript_jni,Unknown,RS compat mode
2020-05-14 02:35:02,VERBOSE,RenderScript_jni,Unknown,Unable to load libRSSupportIO.so, USAGE_IO not supported
2020-05-14 02:35:02,ERROR,linker,Unknown,library "/system/lib64/libblasV8.so" ("/system/lib64/libblasV8.so") needed or dlopened by "/system/lib64/libnativeloader.so" is not accessible for the namespace: [name="classloader-namespace", ld_library_paths="", default_library_paths="/data/app/de.markusfisch.android.binaryeye-2/lib/arm64:/data/app/de.markusfisch.android.binaryeye-2/base.apk!/lib/arm64-v8a", permitted_paths="/data:/mnt/expand:/data/data/de.markusfisch.android.binaryeye"]
2020-05-14 02:35:02,VERBOSE,RenderScript_jni,Unknown,Unable to load BLAS lib, ONLY BNNM will be supported: java.lang.UnsatisfiedLinkError: dlopen failed: library "/system/lib64/libblasV8.so" needed or dlopened by "/system/lib64/libnativeloader.so" is not accessible for the namespace "classloader-namespace"
2020-05-14 02:35:02,ERROR,RenderScript,Unknown,Couldn't load libRSSupportIO.so
2020-05-14 02:35:02,ERROR,RenderScript,Unknown,Unable to open shared library (/data/app/de.markusfisch.android.binaryeye-2/lib/arm64/librs.rotator.so): dlopen failed: invalid ELF file "/data/app/de.markusfisch.android.binaryeye-2/lib/arm64/librs.rotator.so" load segment[4]: p_offset (0x30000) + p_filesz (0x0) ( = 0x30000) past end of file (0x20688)
2020-05-14 02:35:02,ERROR,RenderScript,Unknown,Unable to open system shared library (/system/lib/librs.rotator.so): (null)
```
Full log attached.
[logcat-binaryeye-20200514.csv.log](https://github.com/markusfisch/BinaryEye/files/4628449/logcat-binaryeye-20200514.csv.log)

**Comment by markusfisch** [comment]
Please try Version [1.28.0](https://github.com/markusfisch/BinaryEye/releases/tag/1.28.0) either directly from GitHub:
https://github.com/markusfisch/BinaryEye/releases/tag/1.28.0

Or by becoming a beta tester on Google Play (don't worry, you can go back to production at any time):
https://play.google.com/apps/testing/de.markusfisch.android.binaryeye

Or until the version hits FDroid 😉 What is going to happen pretty soon, probably.

**Comment by jjakob** [comment]
1.28.0 works for me :) thanks

**Comment by markusfisch** [comment]
Great! 😄 Thanks a lot for confirming it! 👍

**Comment by arnep-de** [comment]
1.28.0 works for me, too: Test OK ✅
